src/Engine/ProtoCore/RuntimeStatus.cs

```diff
@@ -70,35 +72,50 @@ public IOutputStream WebMessageHandler
             set;
         }
 
-        public IEnumerable<Runtime.WarningEntry> Warnings
+        public List<Runtime.WarningEntry> Warnings
         {
             get
             {
-                return warnings;
+                using (rwl.CreateReadLock())
+                {
+                    return warnings.ToList();
```

Comment:
this is the slowest part.
There is no good way to ensure thread safety for a Linq deferred execution/query.
One other way would be to use more granular locking ... like the System.Collections.Concurrent Namespace

Reply:
At the moment, Dynamo calls  the `Warnings` getter is only at the end of each Graph execution.
I t should not be that bad of a perf degradation